src/mlpack/core/data/csv_parser.hpp

```diff
@@ -0,0 +1,105 @@
+
+#include <limits>
+
+namespace mlpack{
+namespace data{
+
+enum struct file_type : unsigned int
+{
+  mlp_file_type_unknown,
+  mlp_auto_detect,        //!< attempt to automatically detect the file type
+  mlp_raw_ascii,          //!< raw text (ASCII), without a header
+  mlp_arma_ascii,         //!< Armadillo text format, with a header specifying matrix type and size
+  mlp_csv_ascii,          //!< comma separated values (CSV), without a header
+  mlp_raw_binary,         //!< raw binary format (machine dependent), without a header
+  mlp_arma_binary,        //!< Armadillo binary format (machine dependent), with a header specifying matrix type and size
+  mlp_pgm_binary,         //!< Portable Grey Map (greyscale image)
+  mlp_ppm_binary,         //!< Portable Pixel Map (colour image), used by the field and cube classes
+  mlp_hdf5_binary,        //!< HDF5: open binary format, not specific to Armadillo, which can store arbitrary data
+  mlp_hdf5_binary_trans,  //!< [DO NOT US deprecated] as per hdf5_binary, but save/load the data with columns transposed to rows
+  mlp_coord_ascii         //!< simple co-ordinate format for sparse matrices (indices start at zero)
+  };
+
+/**
+ * WHere should I place this fucntion?
+ * This fucntion is used to convert mlpack file type to respective
+ * arma file type.
+ *
+ * @param type Mlpack's file_type which will we converted to arma's file_type
+ */
+arma::file_type ToArmaFileType(file_type& type);
+
+/**
+ * Convert the given string token to assigned datatype and assign
+ * this value to the given address. The address here will be a
+ * matrix location.
+ * 
+ * Token is always read as a string, if the given token is +/-INF or NAN
+ * it converts them to infinity and NAN using numeric_limits.
+ *
+ * @param val Token's value will be assigned to this address
+ * @param token Value which should be assigned
+ */
+template<typename MatType>
+bool ConvertToken(typename MatType::elem_type& val, const std::string& token);
+
+/**
+ * Returns a bool value showing whether data was loaded successfully or not.
+ *
+ * Parses the file and loads the data into the given matrix. It will make the
+ * first parse to determine the number of cols and rows in the given file.
+ * Once the rows and cols are fixed we initialize a matrix of size(which we
+ * calculated in the first pass) and fill it with zeros. In the second pass
+ * it converts each value to required datatype and sets it equal to val.
+ * 
+ * Using MatType as template parameter here so that in future if mlpack
+ * decides to use any other linear algebra library or want to support
+ * multiple linear algebra libraries, we can make the transition easily.
+ * This is to make the csv parser as generic as possible.
+ * 
+ * @param x Matrix in which data will be loaded
+ * @param f File stream to access the data file
+ */
+template<typename MatType>
+bool LoadCSVV(MatType& x, std::fstream& f);
+
+} // namespace data
+} // namespace mlpack
+
+#endif
```

Comment:
@heisenbuug You have to add `#include csv_parser_impl.hh` at line 104, Since it the implementation is header only so it needs to be included in this file too.